pkg/netpol/eval/check_eval.go

```diff
@@ -0,0 +1,257 @@
+	if hasResult {
+		return res, nil
+	}
+
+	egressRes, err := pe.allowedXgressConnection(srcPeer, dstPeer, false, protocol, port)
+	if err != nil {
+		return false, err
+	}
+	if !egressRes {
+		pe.cache.addConnectionResult(srcPeer, dstPeer, protocol, port, false)
+		return false, nil
+	}
+	ingressRes, err := pe.allowedXgressConnection(srcPeer, dstPeer, true, protocol, port)
+	if err != nil {
+		return false, err
+	}
+	pe.cache.addConnectionResult(srcPeer, dstPeer, protocol, port, ingressRes)
+	return ingressRes, nil
+}
+
+// allowedXgressConnection returns if the given input connection is allowed on the given ingress/egress direction
+// by k8s policies api
+func (pe *PolicyEngine) allowedXgressConnection(src, dst k8s.Peer, isIngress bool, protocol, port string) (bool, error) {
+	// first checks if the connection is allowed or denied by the admin-network-policies (return anpRes).
+	// if the connection is passed by the ANPs or not captured by them, then will continue to NPs (pass = true)
+	anpRes, pass, err := pe.allowedXgressConnectionByAdminNetpols(src, dst, isIngress, protocol, port)
+	if err != nil {
+		return false, err
+	}
+	if !pass { // i.e the connection is captured by the adminNetworkPolicies and definitely is either allowed or denied
+		pe.cache.addConnectionResult(src, dst, protocol, port, anpRes)
+		return anpRes, nil
+	}
+	// else pass == true : means that:
+	// - the admin-network-policies did not capture the connection (if the ANPs rules did not capture the connection explicitly,
+	// then it is not captured.)
+	// - or that the rules captured the connection with action: pass; so it will be determined with netpols/ banp.
+	netpolRes, captured, err := pe.allowedXgressConnectionByNetpols(src, dst, isIngress, protocol, port)
+	if err != nil {
+		return false, err
+	}
+	// if the src/dst was captured by the relevant xgress policies, then the connection is
+	// definitely allowed or denied by the policy rules (either explicitly or implicitly)
+	if captured {
+		pe.cache.addConnectionResult(src, dst, protocol, port, netpolRes)
+		return netpolRes, nil
+	}
+	// else !captured : means that the xgress connection will be determined by the baseline-admin-network-policy,
+	// or the system-default for connection that was not captured by any policy which is allowed.
+	defaultRes, err := pe.allowedXgressByBaselineAdminNetpolOrByDefault(src, dst, isIngress, protocol, port)
+	if err != nil {
+		return false, err
+	}
+	pe.cache.addConnectionResult(src, dst, protocol, port, defaultRes)
+	return defaultRes, nil
+}
+
+// allowedXgressConnectionByAdminNetpols returns if the given input connection is allowed on the given ingress/egress direction
+// by k8s admin-network-policies
+func (pe *PolicyEngine) allowedXgressConnectionByAdminNetpols(src, dst k8s.Peer, isIngress bool, protocol, port string) (res, pass bool,
+	err error) {
+	// iterate sorted by priority admin netpols
+	for _, anp := range pe.sortedAdminNetpols {
+		if isIngress {
```

Comment:
admin netpols are cluster scoped - don't have namespace

Reply:
maybe divide by the subject namespaces?

Reply:
subject may be by namespaces or by pods (and a pod name may match more than one namespace).
I don't think this would be more efficient than what happens now;
 we'll need to loop the admin-netpols after inserting all namespaces and workloads and then parse their subjects in order to store in a map;